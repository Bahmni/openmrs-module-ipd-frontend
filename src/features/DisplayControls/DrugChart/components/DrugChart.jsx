import React, { useEffect, useRef } from "react";
import PropTypes from "prop-types";
import useScrollSync from "react-scroll-sync-hook";
import Calendar from "./Calendar";
import CalendarHeader from "./CalendarHeader";
import "../styles/DrugChart.scss";
import DrugList from "./DrugList";
import DrugChartLegend from "../../../../components/AdministrationLegend/AdministrationLegend";
export default function DrugChart(props) {
  const { drugChartData, currentShiftArray, selectedDate } = props;
  const leftPane = useRef(null);
  const rightPane = useRef(null);
  const { registerPane, unregisterPane } = useScrollSync({
    vertical: true,
  });
  const drugChartRowHeight = 66;
  const drugChartHeight = (drugChartData[0].length + 2) * drugChartRowHeight;
  useEffect(() => {
    if (leftPane.current) {
      registerPane(leftPane.current);
    }
    if (rightPane.current) {
      registerPane(rightPane.current);
    }
    return () => {
      if (leftPane.current) {
        unregisterPane(leftPane.current);
      }
      if (rightPane.current) {
        unregisterPane(rightPane.current);
      }
    };
  }, [leftPane, rightPane, registerPane, unregisterPane]);
  return (
    <div className="drug-chart-dashboard">
      <div className="drug-chart" style={{ height: drugChartHeight }}>
        <div
          className="drug-chart-left-panel"
          data-testid="left-panel"
          ref={leftPane}
        >
          <div className="header" />
          <DrugList drugDetails={drugChartData[1]} />
        </div>
        <div
          className="drug-chart-content"
          data-testid="right-panel"
          ref={rightPane}
        >
          <CalendarHeader currentShiftArray={currentShiftArray} />
          <Calendar
            calendarData={drugChartData[0]}
            currentShiftArray={currentShiftArray}
            selectedDate={selectedDate}
          />
        </div>
      </div>
      <DrugChartLegend />
    </div>
  );
}
DrugChart.propTypes = {
  drugChartData: PropTypes.array.isRequired,
  currentShiftArray: PropTypes.array,
  selectedDate: PropTypes.instanceOf(Date),
};
