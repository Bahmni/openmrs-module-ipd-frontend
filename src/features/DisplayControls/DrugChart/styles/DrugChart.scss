.drug-chart-dashboard {
  font-family: "IBM Plex Sans", sans-serif;
}
.drug-chart {
  --time-cell-height: 66px;
  --half-hour-width: 21px;
  --line-width: 1px;

  max-height: calc(8 * var(--time-cell-height));
  display: grid;
  grid-template-columns: 3fr 7fr;
  width: 100%;
  border: 1px solid #ccc;
  margin: 0 auto;
}

.drug-chart-left-panel {
  border-right: 1px solid #ccc;
  font-size: 13px;
  overflow-y: auto;
  height: 100%;
  min-width: 225px;
  box-shadow: 0px 1px 4px 2px rgba(0, 0, 0, 0.25);
  z-index: 20;
  .header {
    height: 63px;
    background-color: #ededed;
    position: sticky;
    top: 0;
  }
  table {
    table-layout: fixed;
    border-spacing: 0;
    width: 100%;
    tbody > tr:nth-child(even) {
      background-color: #f4f4f4;
    }
    tr {
      height: 66px;
      border-bottom: 0;
      td {
        line-height: 20px;
        padding: 0 10px;
        max-width: 100%;
        border: 0;
      }
    }
  }
  -ms-overflow-style: none; /* IE and Edge */
  scrollbar-width: none; /* Firefox */
}

.drug-chart-left-panel::-webkit-scrollbar {
  display: none;
}

.drug-chart-content {
  overflow: auto;
  height: 100%;
}
