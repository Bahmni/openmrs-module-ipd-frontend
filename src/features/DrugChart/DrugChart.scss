.drug-chart-dashboard {
  font-family: "IBM Plex Sans", sans-serif;
}
.drug-chart {
  --time-cell-height: 66px;
  --half-hour-width: 21px;
  --line-width: 1px;

  height: calc(
    7 * var(--time-cell-height)
  ); //TODO: Update the fixed height once the full dashboard is completed
  display: grid;
  grid-template-columns: 3fr 7fr;
  width: 100%;
  border: 1px solid #ccc;
  margin: 0 auto;
}

.drug-chart-left-panel {
  border-right: 1px solid #ccc;
  font-size: 13px;
  overflow-y: auto;
  height: 100%;
  min-width: 225px;
  .header {
    height: 63px;
    background-color: #ededed;
    position: sticky;
    top: 0;
  }
  table {
    table-layout: fixed;
    border-spacing: 0;
    width: 100%;
    tbody > tr:nth-child(even) {
      background-color: #f4f4f4;
    }
    tr {
      height: 66px;
      border-bottom: 0;
      td {
        line-height: 20px;
        padding: 0 10px;
        max-width: 100%;
        border: 0;
      }
    }
  }
}

.drug-chart-content {
  overflow: auto;
  height: 100%;
}
