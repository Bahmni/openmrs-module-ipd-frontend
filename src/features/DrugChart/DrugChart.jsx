import React from "react";
import PropTypes from "prop-types";
import moment from "moment";
import Calendar from "../Calendar/Calendar";
import CalendarHeader from "../CalendarHeader/CalendarHeader";
import "./DrugChart.scss";

export const TransformDrugChartData = (drugChartData) => {
  const transformedDrugChartData = drugChartData.map((schedule) => {
    const { slots } = schedule;
    const slotData = {};
    slots.forEach((slot) => {
      const { startDateTime, status } = slot;
      const startDateTimeObj = new Date(startDateTime);
      let adminInfo;
      if (slot.admin) {
        const { administeredBy, administeredAt } = slot.admin;
        const administeredTime = moment(new Date(administeredAt)).format(
          "HH:mm"
        );

        adminInfo = administeredBy + " [" + administeredTime + "]";
      }

      const startHour = startDateTimeObj.getHours();
      const startMinutes = startDateTimeObj.getMinutes();
      slotData[startHour] = {
        minutes: startMinutes,
        status: status,
        administrationInfo: adminInfo,
      };
    });
    return slotData;
  });
  return transformedDrugChartData;
};

export default function DrugChart(props) {
  const { patientId, viewDate = new Date() } = props;
  console.log("DrugChart: patientId: ", patientId);
  console.log("DrugChart: viewDate: ", viewDate);

  const drugChartData = [
    {
      scheduleId: 1,
      scheduleUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
      scheduleServiceType: "Medication Administration",
      patientUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
      comment: "some comment",
      startDate: "2023-08-08T18:30:00.000Z",
      endDate: "2023-08-08T18:30:00.000Z",
      slots: [
        {
          id: 1,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 11,
          serviceType: "Medication Administration",
          status: "Not-Administered",
          startDateTime: "2023-08-08T08:30:00.000Z",
          endDateTime: "2023-08-08T09:30:00.000Z",
          notes: "some slot text",
        },
        {
          id: 2,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 12,
          serviceType: "Medication Administration",
          status: "Administered",
          startDateTime: "2023-08-08T11:30:00.000Z",
          endDateTime: "2023-08-08T09:30:00.000Z",
          notes: "some slot text",
          admin: {
            administeredBy: "Dr. John Doe",
            administeredAt: "2023-08-08T08:30:00.000Z",
            adminid: "1234",
          },
        },
        {
          id: 3,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 12,
          serviceType: "Medication Administration",
          status: "Pending",
          startDateTime: "2023-08-08T14:30:00.000Z",
          endDateTime: "2023-08-08T09:30:00.000Z",
          notes: "some slot text",
        },
      ],
    },
    {
      scheduleId: 2,
      scheduleUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
      patientUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
      scheduleServiceType: "Medication Administration",
      comment: "some comment",
      scheduleStartDate: "2023-08-08T18:30:00.000Z",
      scheduleEndDate: "2023-08-08T18:30:00.000Z",
      slots: [
        {
          id: 3,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 11,
          serviceType: "Medication Administration",
          status: "Administered-Late",
          startDateTime: "2023-08-08T08:30:00.000Z",
          endDateTime: "2023-08-08T09:30:00.000Z",
          notes: "some slot text",
          admin: {
            administeredBy: "Dr. John Doe",
            administeredAt: "2023-08-08T08:30:00.000Z",
            adminid: "1234",
          },
        },
        {
          id: 4,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 12,
          serviceType: "Medication Administration",
          status: "Late",
          startDateTime: "2023-08-08T02:30:00.000Z",
          endDateTime: "2023-08-08T03:30:00.000Z",
          notes: "some slot text",
          admin: {
            administeredBy: "Dr. John Doe",
            administeredAt: "2023-08-08T08:30:00.000Z",
            adminid: "1234",
          },
        },
      ],
    },
    {
      scheduleId: 3,
      scheduleUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
      patientUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
      scheduleServiceType: "Medication Administration",
      comment: "some comment",
      scheduleStartDate: "2023-08-08T18:30:00.000Z",
      scheduleEndDate: "2023-08-08T18:30:00.000Z",
      slots: [
        {
          id: 3,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 11,
          serviceType: "Medication Administration",
          status: "Administered",
          startDateTime: "2023-08-08T06:30:00.000Z",
          endDateTime: "2023-08-08T09:30:00.000Z",
          notes: "some slot text",
          admin: {
            administeredBy: "Dr. John Doe",
            administeredAt: "2023-08-08T06:30:00.000Z",
            adminid: "1234",
          },
        },
        {
          id: 4,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 12,
          serviceType: "Medication Administration",
          status: "Late",
          startDateTime: "2023-08-08T04:30:00.000Z",
          endDateTime: "2023-08-08T05:30:00.000Z",
          notes: "some slot text",
        },
      ],
    },
    {
      scheduleId: 4,
      scheduleUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
      scheduleServiceType: "Medication Administration",
      patientUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
      comment: "some comment",
      startDate: "2023-08-08T18:30:00.000Z",
      endDate: "2023-08-08T18:30:00.000Z",
      slots: [
        {
          id: 1,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 13,
          serviceType: "Medication Administration",
          status: "Not-Administered",
          startDateTime: "2023-08-08T10:45:00.000Z",
          endDateTime: "2023-08-08T09:30:00.000Z",
          notes: "some slot text",
        },
        {
          id: 2,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 12,
          serviceType: "Medication Administration",
          status: "Administered",
          startDateTime: "2023-08-08T13:35:00.000Z",
          endDateTime: "2023-08-08T09:30:00.000Z",
          notes: "some slot text",
          admin: {
            administeredBy: "Dr. John Doe",
            administeredAt: "2023-08-08T13:40:00.000Z",
            adminid: "1234",
          },
        },
        {
          id: 3,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 12,
          serviceType: "Medication Administration",
          status: "Pending",
          startDateTime: "2023-08-08T15:30:00.000Z",
          endDateTime: "2023-08-08T09:30:00.000Z",
          notes: "some slot text",
        },
      ],
    },
  ];
  const transformedDrugchartData = TransformDrugChartData(drugChartData);

  return (
    <div className="drug-chart">
      <div className="drug-chart-left-panel"></div>
      <div className="drug-chart-content">
        <CalendarHeader />
        <Calendar calendarData={transformedDrugchartData} />
      </div>
    </div>
  );
}

DrugChart.propTypes = {
  patientId: PropTypes.string.isRequired,
  viewDate: PropTypes.instanceOf(Date),
};
