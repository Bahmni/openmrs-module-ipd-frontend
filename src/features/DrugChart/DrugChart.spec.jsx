import React from "react";
import { render } from "@testing-library/react";
import DrugChart, { TransformDrugChartData } from "./DrugChart";

const mockCalendar = jest.fn();
jest.mock("../CalendarHeader/CalendarHeader", () => {
  return function CalendarHeader() {
    return <div>CalendarHeader</div>;
  };
});

jest.mock("../Calendar/Calendar", () => {
  return function Calendar(props) {
    mockCalendar(props);
    return <div>Calendar</div>;
  };
});

const drugChartData = [
  {
    scheduleId: 1,
    scheduleUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
    scheduleServiceType: "Medication Administration",
    patientUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
    comment: "some comment",
    startDate: "2023-08-08T18:30:00.000Z",
    endDate: "2023-08-08T18:30:00.000Z",
    order: {
      drug: {
        display: "Paracetamol 120 mg/5 mL Suspension (Liquid)",
      },
      route: {
        uuid: "9d6bc13f-3f10-11e4-adec-0800271c1b75",
        display: "Oral",
      },
      dose: 25,
      doseUnits: {
        uuid: "86239663-7b04-4563-b877-d7efc4fe6c46",
        display: "ml",
      },
      duration: 3,
      durationUnits: {
        uuid: "9d7437a9-3f10-11e4-adec-0800271c1b75",
        display: "Day(s)",
      },
    },
    slots: [
      {
        id: 1,
        uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
        orderId: 11,
        serviceType: "Medication Administration",
        status: "Not-Administered",
        startDateTime: "2023-08-08T08:30:00.000Z",
        endDateTime: "2023-08-08T09:30:00.000Z",
        notes: "some slot text",
        admin: {
          administeredBy: "Dr. John Doe",
          administeredAt: "2023-08-08T08:30:00.000Z",
          adminid: "1234",
        },
      },
      {
        id: 2,
        uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
        orderId: 12,
        serviceType: "Medication Administration",
        status: "Administered",
        startDateTime: "2023-08-08T11:30:00.000Z",
        endDateTime: "2023-08-08T09:30:00.000Z",
        notes: "some slot text",
        admin: {
          administeredBy: "Dr. John Doe",
          administeredAt: "2023-08-08T11:40:00.000Z",
          adminid: "1234",
        },
      },
      {
        id: 3,
        uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
        orderId: 12,
        serviceType: "Medication Administration",
        status: "Pending",
        startDateTime: "2023-08-08T15:30:00.000Z",
        endDateTime: "2023-08-08T09:30:00.000Z",
        notes: "some slot text",
        admin: {
          administeredBy: "Dr. John Doe",
          administeredAt: "2023-08-08T15:30:00.000Z",
          adminid: "1234",
        },
      },
    ],
  },
];
describe("DrugChart", () => {
  it("should match snapshot", () => {
    const { asFragment } = render(<DrugChart patientId="test-id" />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("TransformDrugChartData", () => {
  it("should transform drug chart data", () => {
    const TransformedDrugChartData = TransformDrugChartData(drugChartData);
    expect(TransformedDrugChartData).toEqual([
      [
        {
          11: {
            administrationInfo: "Dr. John Doe [11:40]",
            minutes: 30,
            status: "Administered",
          },
          15: {
            administrationInfo: "Dr. John Doe [15:30]",
            minutes: 30,
            status: "Pending",
          },
          8: {
            administrationInfo: "Dr. John Doe [08:30]",
            minutes: 30,
            status: "Not-Administered",
          },
        },
      ],
      [
        {
          administrationInfo: [
            {
              kind: "Administered",
              time: "11:40",
            },
          ],
          dosage: "25ml",
          drugName: "Paracetamol 120 mg/5 mL Suspension (Liquid)",
          drugRoute: "Oral",
          duration: "3 Day(s)",
        },
      ],
    ]);
  });
});
