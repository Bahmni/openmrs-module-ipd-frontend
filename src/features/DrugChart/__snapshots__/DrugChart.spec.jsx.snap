// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrugChart should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="drug-chart"
  >
    <div
      class="drug-chart-left-panel"
    />
    <div
      class="drug-chart-content"
    >
      <div>
        CalendarHeader
      </div>
      <div>
        Calendar
      </div>
    </div>
  </div>
</DocumentFragment>
`;
